File: 621-task-scheduler/task-scheduler.java
class Solution {
    public int leastInterval(char[] tasks, int n) {
        int[] f = new int[26];
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        Queue<int[]> q = new LinkedList<>();
        for(char c : tasks){
            f[c-'A']++;
        }
        for(int i=0;i<26;i++)
        {
            if(f[i]>0) maxHeap.offer(f[i]);
        }
        int time = 0;
        while(!maxHeap.isEmpty() || !q.isEmpty())
        {
            time++;
            if(!maxHeap.isEmpty())
            {
                int freq = maxHeap.poll();
                if(freq>1) q.offer(new int[]{freq-1,time+n});
            }
            else{
                time = q.peek()[1];
            }
            if(!q.isEmpty() && time==q.peek()[1]){
                maxHeap.offer(q.poll()[0]);
            }
            
        }
        return time;
    }
}

File: 621-task-scheduler/task-scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        priority_queue<int> pq;
        vector<int> freq(26,0);//since only alphabets
        for(char ch : tasks) freq[ch-'A']++;
        for(int i : freq) if(i>0) pq.push(i);
        //only consider frequencies greater than 1
        queue<pair<int,int>> q;
        int time = 0;
        while(!pq.empty() || !q.empty())
        {
          time++;
          if(!pq.empty())
          {
            int topElement = pq.top();
            pq.pop();
            if(topElement>1)
            {
                q.push(make_pair(time+n,topElement-1));
            }
          }
          else{
            time = q.front().first;
          }
          if(!q.empty() && time==q.front().first){
            pq.push(q.front().second);
            q.pop();
          }
        }
        return time;
      
    }
};